### PR TITLE
v: support generic interfaces and sumtypes

### LINKED ISSUE #9735: Generics with interfaces
It would be nice to be able to add type parameters to interfaces. This would allow writing generic code without knowing too much in advance about what might be inside containers. For instance,
```v
interface Gettable<T> {
  get() T
}

struct Animal<T> {
  metadata T
}

fn (a Animal) get<T>() T {
  return a.metadata
}

fn extract([]Gettable<T> xs) []T {
  return xs.map(it.get())
}
```

### PR DESCRIPTION
**Pull request by spaceface777**
Closes #9735. (the code in that issue has been added as a functionality test).
May also close other generics-related issues, as I had to (slightly) modify the existing implementation to make this work.

___

This PR implements generic interfaces and sumtypes. This allows for things like:
```v
type Option<T> = T | None | Error
```
or: 
```v
interface Iterator<T> {
    next() ?T
}
```
etc. These two type definitions specifically have been brought up many times in the past over on discord, and are good examples of cases where they can be useful.

___

I know this PR is quite big (and hard to review), but I couldn't really split it into smaller PRs<sup>1</sup>. What I have done, though, is split it into small(er) *commits*, which each add/change/fix one thing, so that it is clear what exactly this PR changes and in what order it should be reviewed<sup>2</sup>.
*TL;DR: please review this PR by commits, and not by looking at the overall diff.*

Also, there are most definitely bugs in these new features. The existing generic types are still not fully stable (although they have improved quite a bit recently, thanks to yuyi's work), and I didn't want to bloat this PR even further with bug fixes. The test files added are quite basic, but they do pass :-). Also, there are still changes here that are untested (especially checker errors) - I will probably add tests for these in future PRs.

___

<sup>1. Interfaces and sumtypes are implemented quite similarly in the compiler right now, and this was also the case for their respective generics implementations. Also, working on one helped me debug the other in some cases.</sup>

<sup>2. Before cleaning up the git history, I had over 40 commits and 2500 changed lines, as I had to go back and e.g. change the checker to fix a codegen issue several times - and _this_ was truly impossible to review. Hopefully the smaller commits help.</sup>

### CODE REVIEW

**Comment by yuyi98** [comment]
@spaceface777   Great job !!!

**Review by yuyi98 — APPROVED** [review]

**Comment by SleepyRoy** [comment]
Good work, finally has this feature!
Is generic struct as variant supported with this PR? For example type Tree<T> = Empty | Node<T>

**Comment by spaceface777** [comment]
@SleepyRoy it does.

The test file here has a `type MyType<T> = Bar<T> | Foo<T>` line, but I've tested the code below and it also works correctly:
```v
struct Leaf<T> {
	val   T
	left  Node<T>
	right Node<T>
}

struct Empty {}

type Node<T> = Leaf<T> | Empty

x := Node<int>(Leaf<int>{
	val: 3
	left: Leaf<int>{
		val: 1
		left: Empty{}
		right: Leaf<int>{
			val: 2
			left: Empty{}
			right: Empty{}
		}
	}
	right: Empty{}
})

println(x)
```

**Comment by SleepyRoy** [comment]
@spaceface777 Fantastic! So I would finally be able to update my BST example with generics after your PR is merged : )

File: vlib/v/parser/parser.v
Comment (spytheman) [inline vlib/v/parser/parser.v:1923]: why 20?

Comment (spytheman) [reply vlib/v/parser/parser.v:1923]: should not there be a `if p.tok.kind == .eof {` check here too?

File: vlib/v/parser/struct.v
Comment (spytheman) [inline vlib/v/parser/struct.v:470]: That is the same as the code in name_expr ... a common function?

File: vlib/v/parser/parser.v
Comment (spaceface777) [reply vlib/v/parser/parser.v:1923]: > why 20?

it's a completely arbitrary limit I put to avoid looking too far ahead, though it may not be needed since we do the `valid_tokens_inside_types` check anyways.

> should not there be a `if tok.kind == .eof {` check here too?

I don't think it's necessary, since `.eof` is already `!in parser.valid_tokens_inside_types` :)

File: vlib/v/parser/struct.v
Comment (spaceface777) [reply vlib/v/parser/struct.v:470]: sure, good idea.

File: vlib/v/tests/generic_interface_test.v
Comment (spytheman) [inline vlib/v/tests/generic_interface_test.v:12]: Can you add one more struct similar to Animal<T>, say Mineral<T>, so that it tests that the generated interface code can work with both?

Comment (spytheman) [reply vlib/v/tests/generic_interface_test.v:12]: Another thing to test is something like `Animal<f32>` in the same test file.

File: vlib/v/tests/generic_sumtype_test.v
Comment (spytheman) [inline vlib/v/tests/generic_sumtype_test.v:4]: wow, that is really powerful ... very nice!

File: vlib/v/parser/struct.v
Comment (spaceface777) [reply vlib/v/parser/struct.v]: actually, it turns out that there was already a method for this

File: vlib/v/token/token.v
Comment (spytheman) [inline vlib/v/token/token.v:484]: good find

File: vlib/v/tests/generic_interface_test.v
Comment (spaceface777) [reply vlib/v/tests/generic_interface_test.v:12]: I've added a test for the former; the latter doesn't work right now though (there's a type inference bug in the checker). 

I can (try to) fix it in this PR, or I can fix it in a future PR when this is merged - it depends on whether you think I should keep adding things here.

Comment (spytheman) [reply vlib/v/tests/generic_interface_test.v:12]: No, a separate PR will be better I think.

Comment (spaceface777) [reply vlib/v/tests/generic_interface_test.v:12]: ... so i just spent the past 30 minutes trying to debug this, and it turns out it isn't a problem with this PR, but with the existing generics :|

I've opened a bug report for it: #10800.

@yuyi98 ^

Comment (spytheman) [reply vlib/v/tests/generic_interface_test.v:12]: probably related:
```v
interface Gettable<T> { get() T }

struct Animal<T> { 	metadata T }
fn (a Animal<T>) get<T>() T { return a.metadata }

struct Mineral<T> { value T }
fn (m Mineral<T>) get<T>() T { return m.value }

fn extract<T>(xs []Gettable<T>) []T { return xs.map(it.get()) }
fn extract_basic<T>(xs Gettable<T>) T { return xs.get() }

fn test_extract_multiple_instance_types() {
 	mut c := Animal<int>{314}
 	// mut m := Mineral<int>{314} // uncommenting this line fixes the compilation error, even though m is not used anywhere
	x := Gettable<int>(c)
	println(x)
}
```

Comment (spytheman) [reply vlib/v/tests/generic_interface_test.v:12]: with the line commented:
```
x_test.v:7:41: error: cannot use `void` as type `int` in return argument
    5 | 
    6 | struct Mineral<T> { value T }
    7 | fn (m Mineral<T>) get<T>() T { return m.value }
      |                                         ~~~~~
```

Comment (spaceface777) [reply vlib/v/tests/generic_interface_test.v:12]: ah, no - that's unrelated to the above, but it's the same problem as the one i described in this comment in the sumtype test (https://github.com/vlang/v/pull/10795/files#diff-6d6a77193bcb8a0a53c50d83335dd49320afa37ad292659af81d2e7f68ed75f1R51)

Basically, the problem is that, with that line commented, the variant type symbol is not resolved (it does not exist in `Table`, so attempting to get any of its fields' types results in `void`.

**Comment by dumblob** [comment]
A lot of great work! Appreciated!

Could also the following example be modified? I find it highly misleading in that it gives impression that `T | None | Error` is a *V optional* which it isn't.
```
type Option<T> = T | None | Error
```
I can see 2 problems - namely `Option<>` might lead people to think [V has reified optionals](https://github.com/vlang/v/issues/9598 ) and second `Error` might be easily confused with `IError` which is forbidden to be used in sum types as discussed few days ago in another issue/PR I couldn't quickly find now (to hold up to the "there is only one way of doing a thing in V" - because `IError` is meant to be used with `?` i.e. true V optionals).

Maybe turn it into:
```
type MyOption<T> = T | None | MyError
```

**Comment by JalonSolov** [comment]
Wouldn't `MyGeneric<T>` be even better, so there is no mention of option at all?

**Comment by mcastorina** [comment]
Thank you so much for the excellent work @spaceface777 ! This feature is such a great advancement in V's expressiveness as a language!

Regarding this example:
```v
interface Iterator<T> {
    next() ?T
}
```

Were you able to create a struct that implements the interface? My attempts results in C errors, which I can open an issue for, but I I just couldn't figure it out.

**Comment by JalonSolov** [comment]
I thought the "fix" for that (at least for now) was to do
```v
interface Iterator<T> {
    next<T>() ?T
}
```
In other words... for some reason the compiler isn't inferring the correct type.

**Comment by spaceface777** [comment]
@mcastorina this works, for example:
```v
interface Iterator<T> {
	next() ?T
}

struct NumberIterator<T> {
	limit T
mut:
	val T
}

fn (mut i NumberIterator<T>) next<T>() ?T {
	if i.val >= i.limit {
		return none
	}
	i.val++
	return i.val
}

fn main() {
	mut i := Iterator<int>(NumberIterator<int>{ limit: 10 })
	for {
		if val := i.next() {
			println(val)
		} else {
			println('iterator is done')
			break
		}
	}
}
```

Please note that this PR only contains an initial implementation, so bugs are expected (especially given that V's generic structs and functions are still not fully stable ). Thus, more complex/different cases may not work correctly. If you find any such issues, please report them so they can be fixed. 🙂

**Comment by mcastorina** [comment]
Will do! Perhaps there was a regression because that example does not work on V 0.2.2 aca66d5 :confused: 

https://github.com/vlang/v/issues/10815

**Pull request merged**
